Implementation of Stable Diffusion with Aesthetic Gradients

MalumaDev opened the pull request:
here the original repo: https://github.com/vicgalle/stable-diffusion-aesthetic-gradients


![Immagine 2022-10-14 105447](https://user-images.githubusercontent.com/12870488/195807085-8d4897d7-2b76-45a8-8614-4b6242aae7ab.png)
![Immagine 2022-10-14 105513](https://user-images.githubusercontent.com/12870488/195807088-8c86c536-d25e-4be1-907d-384b6bf45551.png)

MalumaDev committed:
init

ShadowPower commented:
```
File "D:\stable-diffusion-webui-aesthetic\modules\sd_hijack.py", line 411, in forward
    z = z * (1 - self.aesthetic_weight) + zn * self.aesthetic_weight
RuntimeError: The size of tensor a (154) must match the size of tensor b (77) at non-singleton dimension 1
```
It seems that the token length is limited by the CLIP model.

EliEron commented:
This seems to work well, but the default values are a bit odd. 

The repo recommends an aesthetic learning rate of 0.0001, but you default to 0.005 which is an order of magnitude higher. Is there a specific reason for this?

Similarly for aesthetic steps the repo recommends starting with relatively small step amounts, but the default in this PR is the highest value that the UI allows.

MalumaDev commented:
To be quick, I put "random" default values 😅
I fixed the problem of the token length and I added the UI for the generation of the embedding. I need some hours of sleep, tomorrow I'll commit the code

MalumaDev committed:
fix to tokens lenght, addend embs generator, add new features to edit the embedding before the generation using text

MalumaDev committed:
Merge branch 'master' into test_resolve_conflicts

vicgalle reviewed (commented):
Thanks for adapting this, @MalumaDev! Looks good to me. 
I only add some suggestions regarding the name of some parameters and the max value of one.

MalumaDev commented on modules/ui.py:
```diff
@@ -592,6 +596,18 @@ def create_ui(wrap_gradio_gpu_call):
+                with gr.Group():
+                    aesthetic_lr = gr.Textbox(label='Learning rate', placeholder="Learning rate", value="0.0001")
+                    aesthetic_weight = gr.Slider(minimum=0, maximum=1, step=0.01, label="Aesthetic weight", value=0.9)
+                    aesthetic_steps = gr.Slider(minimum=0, maximum=256, step=1, label="Aesthetic steps", value=5)
```
Increasing the steps lead to better results than increasing the learning rate. In my test the use of 50 steps lead to interesting results.

EliEron commented on modules/aesthetic_clip.py:
```diff
@@ -0,0 +1,78 @@
+
+
+def check_is_valid_image_file(filename):
+    return filename.lower().endswith(('.png', '.jpg', '.jpeg'))
```
I'd recommend adding WebP and BMP to the list of image format to match the formats allowed by the TI/Hypernetworks training tab.

MalumaDev committed:
Update modules/ui.py

Co-authored-by: Víctor Gallego <vicgalle@ucm.es>

MalumaDev committed:
Update README.md

Co-authored-by: Víctor Gallego <vicgalle@ucm.es>

MalumaDev committed:
Update modules/aesthetic_clip.py

Co-authored-by: Víctor Gallego <vicgalle@ucm.es>

MalumaDev committed:
Update modules/ui.py

Co-authored-by: Víctor Gallego <vicgalle@ucm.es>

MalumaDev committed:
Update modules/ui.py

Co-authored-by: Víctor Gallego <vicgalle@ucm.es>

MalumaDev committed:
Update modules/ui.py

Co-authored-by: Víctor Gallego <vicgalle@ucm.es>

bmaltais commented:
This feature is actually way more interesting than I thought. Pretty amazing the variations you can obtain using the images embeddings. I am still trying to figure out how to use all the different sliders and what they do... I really hope this will get merged someday

bmaltais commented:
Quick example for those wondering. I created an image embedding from a bunch of big eyes paintings and tried to apply it to the simple "a beautiful woman" seed 0 prompt. Here are the results:

Original prompt image:
![image](https://user-images.githubusercontent.com/7474674/195999128-8842d10f-d761-4ec5-a454-ee9ae8239203.png)

Applying the image embedding style with aesthetic: learning rate 0.001, weight 0.85 and steps 40:
![image](https://user-images.githubusercontent.com/7474674/195999187-b759cf4a-054e-4bdc-849c-1e06dc12ad03.png)

Increasing the weight to 0.95 increasing the style application resulting in something closer to the original paintings:
![image](https://user-images.githubusercontent.com/7474674/195999228-c61241b5-ed99-40d2-aaa3-ac5be8618faf.png)

Bringing it down to 50 will obviously reduce:
![image](https://user-images.githubusercontent.com/7474674/195999258-cde0bc43-ef43-490d-bbaf-fdb28877ff6d.png)

bmaltais commented:
Another example using the same prompt as above. I created an image embedding from a bunch of images at: https://lexica.art/?q=aadb4a24-2469-47d8-9497-cafc1f513071

After some fine tuning of the weights and learning rate I was able to get:
![image](https://user-images.githubusercontent.com/7474674/196000138-08adabff-3353-4e60-98f8-121f5d284482.png)

And all this with literally no training what so ever. AMAZING!

MalumaDev commented:
> This feature is actually way more interesting than I thought. Pretty amazing the variations you can obtain using the images embeddings. I am still trying to figure out how to use all the different sliders and what they do... I really hope this will get merged someday.
> 
> I notices creating a new image embedding does not automatically get added to the pull down in text2img. Just a nit pick.

Little bug. I'll fix it.

bmaltais commented:
I even tried feeding it a few pictures of me in a non 1:1 aspect ratio and gosh darn... if produced passable results!

Sample input image:

![00000-0-a man with a beard and a white shirt is smiling at the camera with a waterfall in the background](https://user-images.githubusercontent.com/7474674/196001522-824e47ea-444b-4628-9caa-e23bc78d0251.png)

![image](https://user-images.githubusercontent.com/7474674/196001558-f5c9138d-f5fb-435d-b049-fe48e7aeea20.png)

Not as good as trained Dreambooth or TI but for a 1 minute fiddling it is amazing.

MalumaDev committed:
Add support to other img format, fixed dropbox update

MalumaDev committed:
Merge branch 'master' into test_resolve_conflicts

MalumaDev committed:
fixed dropbox update

MalumaDev committed:
Merge remote-tracking branch 'origin/test_resolve_conflicts' into test_resolve_conflicts

feffy380 commented:
I'd suggest hiding the interface behind the Extra checkbox or similar. It's quite large and pushes more commonly used options like CFG and Batch size/count off-screen.

bmaltais commented:
> I'd suggest hiding the interface behind the Extra checkbox or at least moving it lower. It's quite large and pushes more commonly used options like CFG and Batch size/count off-screen.

Indeed. I doubt Automatic will like it where it is now... best would be some sort of tabs inside the parameter section to present the current options in a default tab and access the aesthetic options in an aesthetic tab beside it.

bmaltais commented:
On a separate note... do you think the same thing could be added to img2img to offer better conformity to the original image? I sometime feel the aesthetic model is difficult to control. A some point it totally change the original image instead of changing the overall style of it. If it was possible to control the weight of the aesthetic on top of the resulting prompt image without it without losing the whole look it would be even better.

bmaltais commented:
Another quick test. Old bearded man:

Prompt no aesthetic:

![00878-2002862293-morgan freeman starring as gandalf in lord of the rings, epic dark fantasy horror stylized oil painting by ivan shiskin  trendin](https://user-images.githubusercontent.com/7474674/196014414-b98d3c76-5422-4547-9180-3638cc4283ab.png)

Aesthetic applied:

![00876-3351746598-morgan freeman starring as gandalf in lord of the rings, epic dark fantasy horror stylized oil painting by ivan shiskin  trendin](https://user-images.githubusercontent.com/7474674/196014421-529e62c6-af06-4d1c-beda-c5f60932a0ee.png)

MalumaDev commented:
> On a separate note... do you think the same thing could be added to img2img to offer better conformity to the original image? I sometime feel the aesthetic model is difficult to control. A some point it totally change the original image instead of changing the overall style of it. If it was possible to control the weight of the aesthetic on top of the resulting prompt image without it without losing the whole look it would be even better.

Yes, I think that it could work right now but I have not added the UI.

MalumaDev commented:
> I'd suggest hiding the interface behind the Extra checkbox or at least moving it lower. It's quite large and pushes more commonly used options like CFG and Batch size/count off-screen.

Today, I will move the board and compact the interface. I think it makes more sense near the prompts, but it can go back there again in the future.

AUTOMATIC1111 commented:
While this really is nice work, I definitely will not accept code that clutters the UI for users who don't want to use this, and I won't accept changes in code where you just take an existing line and change formatting of it without changing what it does.

Changing PIL.Image.BICUBIC to PIL.Image.Resampling.BICUBIC will break some old version of PIL on colabs so do not do that.

Why is there another CLIP being created when we already have one? If it is really needed for this, why is it always created regardless of whether the user wants the gradients?

AUTOMATIC1111 commented:
An additional thing I'm going to ask of you is to isolate as much of your code into separate files as possible. The big chunk of code in sd_hijack should be in its own file. All the parameters of aesthetic gradients should be in members of your own class defined in your own file, not in sd_hijack.

One possible solution for non-cluttered UI is to let use specify an aesthetic embedding as text in the prompt; something like this:

`a tree <aesthetic:weight=0.8, steps=30, slerp>`

This will also have the benefit of putting all paramters into infotext, so that other users you share your prompt with will be able to reproduce it if they have the embedding.

MalumaDev commented:
> While this really is nice work, I definitely will not accept code that clutters the UI for users who don't want to use this, and I won't accept changes in code where you just take an existing line and change formatting of it without changing what it does.
> 
> Changing PIL.Image.BICUBIC to PIL.Image.Resampling.BICUBIC will break some old version of PIL on colabs so do not do that.
> 
> Why is there another CLIP being created when we already have one? If it is really needed for this, why is it always created regardless of whether the user wants the gradients?

The only clip that I found is the CLIPTextModel, but we need also the text_projection that is in the CLIPModel class and the CLIPVisualModel to generate the embs.

MalumaDev committed:
ui fix

MalumaDev committed:
ui fix

MalumaDev commented:
> An additional thing I'm going to ask of you is to isolate as much of your code into separate files as possible. The big chunk of code in sd_hijack should be in its own file. All the parameters of aesthetic gradients should be in members of your own class defined in your own file, not in sd_hijack.
> 
WIP!!

MalumaDev committed:
ui fix, re organization of the code

MalumaDev committed:
Merge branch 'master' into test_resolve_conflicts

MalumaDev commented:
> On a separate note... do you think the same thing could be added to img2img to offer better conformity to the original image? I sometime feel the aesthetic model is difficult to control. A some point it totally change the original image instead of changing the overall style of it. If it was possible to control the weight of the aesthetic on top of the resulting prompt image without it without losing the whole look it would be even better.

Done

MalumaDev committed:
Merge branch 'master' into test_resolve_conflicts

bmaltais commented:
I like the now expandable section for the aesthetic section. This is a step in the right direction and I hope Automatic will approve of it.

I tested the img2img implementation and it work very well. I was able to keep the general composition of the ofiginal and transform it toward the aesthetic without losing too much... NICE. Here is an example of applying the Big Eyes style to a man photo:

Original:

![image](https://user-images.githubusercontent.com/7474674/196055495-92587af2-5621-4b18-b09a-f93aa5e5f84b.png)

Styled with big eyes:

![image](https://user-images.githubusercontent.com/7474674/196055515-14eda7ec-5191-4d7c-a8f6-c472ecb74fff.png)

and the overall config:

![image](https://user-images.githubusercontent.com/7474674/196055539-f470ab7f-b2d8-4ba7-9f71-4b32aa9d6bc3.png)

bmaltais commented:
Something else I noticed. Is there a reason the Aesthetic optimization is always computed? If no parameters for it have changed from generation to generation, could it not just be used from memory cache instead of always being recomputed?

MalumaDev commented:
> Something else I noticed. Is there a reason the Aesthetic optimization is always computed? If no parameters for it have changed from generation to generation, could it not just be used from memory cache instead of always being recomputed?

When the seed changes so does the training result!!!

feffy380 commented:
@bmaltais Looking at the original aesthetic gradients repo, the personalization step involves performing gradient descent to make the prompt embedding more similar to the aesthetic embedding. In other words, it has to be recomputed for each prompt. But it shouldn't be affected by the seed as far as I can tell.

MalumaDev committed:
Merge branch 'master' into test_resolve_conflicts

MalumaDev committed:
Merge branch 'master' into test_resolve_conflicts

MalumaDev committed:
Merge branch 'master' into test_resolve_conflicts

miaw24 commented:
I think there should be an option to do the Aesthetic optimization on cpu, before sending it back to the gpu for the image generation process. This might be useful for people with limited vram, so that they won't run out of vram when computing the Aesthetic optimization

bbecausereasonss commented:
Is there a tutorial on how to set this up/train it?

bmaltais commented:
Have a look over here:  Using Aesthetic Images Embeddings to improve
Dreambooth or TI results · Discussion #3350 ·
AUTOMATIC1111/stable-diffusion-webui (github.com)
<https://github.com/AUTOMATIC1111/stable-diffusion-webui/discussions/3350>

On Fri, Oct 21, 2022 at 11:36 AM becausereasons ***@***.***>
wrote:

> Is there a tutorial on how to set this up/train it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/2585#issuecomment-1287131531>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABZA34T4P2W7UYRGZ3DCM7TWEKZ7HANCNFSM6AAAAAARFBBXIE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

rabidcopy commented:
So is there any hope to do this on 4GB of VRAM? My poor card has been able to handle everything(besides training) up to 576x576 so far with --medvram, VAEs, hypernetworks, upscalers, etc, but this puts me OOM after the first pass. :sweat_smile:

TinyBeeman commented:
It seems like "Aesthetic text for imgs" and slerp angle are somehow off...  Values between 0.001 and 0.02 seem to cause the aesthetic text to influence the embedding in a meaningful way. But 0.2 to 1.0 seem random and not to have that much effect.

MalumaDev commented:
The Dalle2 paper specifies that the max angle to use is in between [0.25,0.5].

TinyBeeman commented:
@MalumaDev that makes sense, maybe we should adjust the slider range to be more helpful. That said, as currently implemented, ranges as low as .001 have interesting variations.

miaw24 commented:
@rabidcopy i was able to use it on 4GB of VRAM by editing `aesthetic_clip.py` where i changed every single `device` to `'cpu'` (except the import part, of course), and then to prevent it from complaining that the tensors are in two different devices, editing the code in `__call__` function of class `AestheticCLIP`, adding `z = z.to('cpu')` before `if self.slerp:` part, and also adding `z = z.to(device)` before the `return z` part. So far, this works (or at least it works for me), but idk whether using cpu to compute the aesthetic gradient will change the result if compared to the result produced while it is computed with cuda.

cornpo commented:
I couldn't run the laion_7plus or sac_8plus since the original pr. Now tonight I can.

hulululu7654321 commented:
File "/home/hulululu/desktop/stable-diffusion-webui-master/extensions/aesthetic-gradients/aesthetic_clip.py", line 233, in call
sim = text_embs @ img_embs.T
RuntimeError: expected scalar type Float but found Half

how can i deal with this problem?

baphilia commented:
> The Dalle2 paper specifies that the max angle to use is in between [0.25,0.5]. (TextDiff)

> It seems like "Aesthetic text for imgs" and slerp angle are somehow off... Values between 0.001 and 0.02 seem to cause the aesthetic text to influence the embedding in a meaningful way. But 0.2 to 1.0 seem random and not to have that much effect relative to each other. If I use "colorful painting", for instance (0.0 = ignore text, 0.001 = it adds color and flowers, 0.2 to 1.0 = the image seems to lose style altogther, and is neither colorful nor painterly.

anyone have a link or a quick explanation of what the 'aesthetic text for imgs', slerp angle, and slerp interpolation are supposed to do?  what should I be typing there? what is the desired effect? 

on low settings for angle it seems to be super random, just changing the entire subject of the image to something that has nothing to do with either the regular prompt or the aesthetic text, and at high settings it just seems to use the aesthetic text as a new prompt (without incorporating the styling of the embedding at all)

jonwong666 commented:
Aesthetic works best with TXT2IMG, its not for IMG2IMG

Im getting good results with these settings 

![image](https://user-images.githubusercontent.com/41051927/199334494-90b17caa-b592-4e44-9d08-409acc1982a0.png)

the trick is to not use too many style or conflicting artists in the main prompt and let aesthetics do the work with a high learning rate.

gsgoldma commented:
> @rabidcopy i am able to use it on 4GB of VRAM by editing `aesthetic_clip.py` where i changed every single `device` to `'cpu'` (except the import part, of course), and then to prevent it from complaining that the tensors are in two different devices, editing the code in `__call__` function of class `AestheticCLIP`, adding `z = z.to('cpu')` before `if self.slerp:` part, and also adding `z = z.to(device)` before the `return z` part. So far, this works (or at least it works for me), but idk whether using cpu to compute the aesthetic gradient will change the result if compared to the result produced while it is computed with cuda.

I might be a fool, but which indentations did you use? I'm getting a cpu is undefined error.

shamblessed reviewed (approved):
Hydd
